### PR TITLE
Hide Console (Rebase of #2082)

### LINKED ISSUE #1669: Add an option to hide the console window
#### Overview:
Add an option to completely hide the console window.

#### Smaller Details: 
A checkbox within options > settings would be a good place for it.

#### Nature of Request:  
- Addition
  - Add a menu checkbox to completely hide the the console window.

### Why would this feature be useful?
This benefits end users who only want to see the main Ryujinx window, as well as those who are using an aggregating emulator frontend (e.g. Hyperspin, Launchbox) and wish for a cleaner experience. 
It would also reduce the number of support questions regarding harmless console output messages that tend to attract attention.

### PR DESCRIPTION
**Pull request by Shahil-Ayato**
This is a rebase of both #2082 and #1866.
The original author of all the code is https://github.com/macabeus
This fixes all the issues the previous PR's had and it fully works on Windows and no longer breaks Linux

Screenshot 2021-05-03 015535

Special thanks to @PixelyIon and @pineappleEA for helping me fix all of the issues with the previous PR's.

Closes #1669.

### CODE REVIEW

File: Ryujinx.Common/Configuration/ConfigurationState.cs [hunk 1]
```diff
@@ -744,6 +752,7 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
             if (configurationFileFormat.Version < 24)
             {
                 Common.Logging.Logger.Warning?.Print(LogClass.Application, $"Outdated configuration version {configurationFileFormat.Version}, migrating to version 24.");
+
```
Comment (AcK77) [inline Ryujinx.Common/Configuration/ConfigurationState.cs:755]: Remove extra empty line

File: Ryujinx.Common/Configuration/ConfigurationState.cs [hunk 2]
```diff
@@ -802,7 +811,16 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
 
                 configurationFileUpdated = true;
             }
+            
+            if (configurationFileFormat.Version < 25)
+            {
+                Common.Logging.Logger.Warning?.Print(LogClass.Application, $"Outdated configuration version {configurationFileFormat.Version}, migrating to version 25.");
 
+                configurationFileFormat.ShowConsole = false;
```
Comment (AcK77) [inline Ryujinx.Common/Configuration/ConfigurationState.cs:819]: Maybe the console should be shown as default as you set it to true in LoadDefault()

File: Ryujinx.Common/System/ConsoleHelper.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                    standardOutput.Dispose();
+
+                    Console.Title = $"Ryujinx Console {Assembly.GetEntryAssembly().GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion}";
+                }
+                else if (GetCurrentProcessId() == GetWindowId())
+                {
+                    ShowWindow(GetConsoleWindow(), SW_SHOW);
+                }
+            }
+            else
```
Comment (AcK77) [inline Ryujinx.Common/System/ConsoleHelper.cs:74]: You can concat the else and the if below

File: Ryujinx/Program.cs [hunk 4]
```diff
@@ -140,6 +138,11 @@ static void Main(string[] args)
                 ConfigurationState.Instance.Ui.StartFullscreen.Value = true;
             }
 
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (AcK77) [inline Ryujinx/Program.cs:141]: This check should be done in the ConsoleHelper class

File: Ryujinx/Ui/Windows/SettingsWindow.cs [hunk 5]
```diff
@@ -441,7 +456,12 @@ private void SaveSettings()
             {
                 ConfigurationState.Instance.System.AudioBackend.Value = (AudioBackend)_audioBackendStore.GetValue(activeIter, 1);
             }
-
+            
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/SettingsWindow.cs:460]: This check should be done in the ConsoleHelper class

**Review by AcK77 — CHANGES_REQUESTED** [review]

**Comment by Shahil-Ayato** [comment]
Will do the rest tomorrow

**Comment by Leischii** [comment]
I'm not sure if this should be merged anytime soon. There are still situations where the emulator freezes and the only way to close it is my closing the console window (or via task manager, but that isn't a good solution imo). Trying to close the emulator window doesn't work in these situations, so until this is fixed, I don't think removing the console is that good. Feel free to disagree.

**Comment by Mou-Ikkai** [comment]
> I'm not sure if this should be merged anytime soon. There are still situations where the emulator freezes and the only way to close it is my closing the console window (or via task manager, but that isn't a good solution imo). Trying to close the emulator window doesn't work in these situations, so until this is fixed, I don't think removing the console is that good. Feel free to disagree.

That's why it's a toggleable option.

**Comment by Zeswen** [comment]
> I'm not sure if this should be merged anytime soon. There are still situations where the emulator freezes and the only way to close it is my closing the console window (or via task manager, but that isn't a good solution imo). Trying to close the emulator window doesn't work in these situations, so until this is fixed, I don't think removing the console is that good. Feel free to disagree.

First of all thank you very much for your efforts and the amazing project @Shahil-Ayato :)

I don't think the window freezing is a real issue here, since Windows handles it via the not responding alert, which access Task Manager internally.

Besides, there is a lot of issues with [Remote Play support](https://www.reddit.com/r/Cuphead/comments/dqobm3) due to the console window instantiation, this feature could really help address it (didn't check the PR yet, will do tomorrow). 

Also, as @Mou-Ikkai mentioned, this is a togglable option enabled by default, so there is not really a huge problem here.    

Please reconsider merging this PR, thanks! :)

**Comment by defunky** [comment]
Any plans on fixing conflicts?

**Comment by Shahil-Ayato** [comment]
> Any plans on fixing conflicts?

I can fix the simple conflicts like config.json and program.cs but j need to talk with the devs on The other two as every time I try to rebase I get errors while building this. I hope to manage to fix this soon.

**Comment by Shahil-Ayato** [comment]
Rebased and ready for review! Apologies for how long this took. 
Thank you to @Sazpaimon for helping me out with the rebase!
P.S Sorry @emmauss I hope this isn't difficult to port to Avalonia.

**Comment by MutantAura** [comment]
Working great on my end. 
I'm unsure what should be default honestly. Perhaps at a later date it may be beneficial to change it to auto-hide. I don't know of any other emulator that has their console out by default but this may be a change once the emulator is more mature.

**Comment by Shahil-Ayato** [comment]
Superseded by #3170.

**Pull request closed without merging**